### PR TITLE
Feature/active cover amount

### PR DESCRIPTION
**Pull request by danoctavian**

### CODE REVIEW

File: contracts/modules/cover/Cover.sol [hunk 1]
```diff
@@ -276,14 +293,14 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon {
 
     uint premiumInPaymentAsset = totalPremiumInNXM * nxmPriceInPayoutAsset / 1e18;
 
-    return (premiumInPaymentAsset, totalPremiumInNXM);
+    return (premiumInPaymentAsset, totalPremiumInNXM, totalCoverAmountInNXM);
```
Comment (horodnicdragos) [inline contracts/modules/cover/Cover.sol:296]: totalCoverAmountInNXM?

File: contracts/modules/cover/Cover.sol [hunk 2]
```diff
@@ -457,6 +490,35 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon {
...
+  /* ========== Active cover amount tracking ========== */
+
+  function updateGlobalActiveCoverAmountPerAsset(uint period, uint amountToCover, uint24 assetId) internal {
+
+    uint activeCoverAmount = getGlobalActiveCoverAmountForAsset(assetId);
+    uint32 currentBucket = SafeUintCast.toUint32(block.timestamp / BUCKET_SIZE);
+
+    globalActiveCoverAmountPerAsset[assetId] = uint96(activeCoverAmount + amountToCover);
+    lastGlobalBuckets[assetId] = currentBucket;
+    globalActiveCoverAmountBucketsPerAsset[assetId][(block.timestamp + period) / BUCKET_SIZE] = uint96(amountToCover);
```
Comment (horodnicdragos) [inline contracts/modules/cover/Cover.sol:502]: Use currendBucket which you have already declared above.

Comment (horodnicdragos) [inline contracts/modules/cover/Cover.sol:502]: Assign (block.timestamp + period) / BUCKET_SIZE to expiryBucket and also use SafeUintCast

File: contracts/modules/cover/Cover.sol [hunk 3]
```diff
@@ -233,18 +250,18 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon {
...
     PoolAllocationRequest[] memory allocationRequests
-  ) internal returns (uint, uint) {
+  ) internal returns (uint, uint totalPremiumInNXM, uint totalCoverAmountInNXM) {
     // convert to NXM amount
     uint nxmPriceInPayoutAsset = pool().getTokenPrice(params.payoutAsset);
-    uint totalPremiumInNXM = 0;
-    uint totalCoverAmountInNXM = 0;
     uint remainderAmountInNXM = 0;
 
     for (uint i = 0; i < allocationRequests.length; i++) {
```
Comment (danoctavian) [inline contracts/modules/cover/Cover.sol:267]: store `.length` before to avoid reads

File: contracts/modules/cover/Cover.sol [hunk 4]
```diff
@@ -351,22 +380,44 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon {
...
-        // retrieve extra required payment
-        retrievePayment(premiumInPaymentAsset - refundInPaymentAsset, buyCoverParams);
-      }
+  function handlePaymentAndRefund(
+    BuyCoverParams memory buyCoverParams,
+    uint totalPremiumInNXM,
+    uint premiumInPaymentAsset,
+    uint refundInNXM
+  ) internal {
+
```
Comment (shark0der) [inline contracts/modules/cover/Cover.sol:396]: you can add an early check for `refundInNXM < totalPremiumInNXM` and exit the function early without performing anything else if we don't need payment.

**Pull request merged**